pyomo/contrib/pynumero/interfaces/pyomo_nlp.py

```diff
@@ -66,9 +69,31 @@ def __init__(self, pyomo_model):
                 elif name[0] == 'c':
                     cdidx[obj()] = int(name[1:])
 
+            # The NL writer advertises the external function libraries
+            # through the PYOMO_AMPLFUNC environment variable.
+            if 'PYOMO_AMPLFUNC' in os.environ:
+                _old_amplfunc = os.environ.get('AMPLFUNC', None)
+                if _old_amplfunc is not None:
+                    os.environ['AMPLFUNC'] += "\n" + os.environ['PYOMO_AMPLFUNC']
+                else:
+                    os.environ['AMPLFUNC'] = os.environ['PYOMO_AMPLFUNC']
+                if os.name in ['nt', 'dos']:
+                    ctypes.cdll.msvcrt._wputenv_s("AMPLFUNC", os.environ["AMPLFUNC"])
+
             # now call the AslNLP with the newly created nl_file
             super(PyomoNLP, self).__init__(nl_file)
 
+            # Restore the AMPLFUNC environment variable
```

Comment:
I think you want to put the `__init__` call in a `try` with the restoration of the environment in a `finally` block

Reply:
Ok, I'm pretty sure I understand, but just to be sure, this is to ensure the environment gets restored if ```__init__``` fails?  Because if someone catches the exception and it doesn't terminate the process, it will mess with my subsequent calls to the nl-writer?

Reply:
Correct.  One of the things that I am trying to get us to be more consistent on is that if we modify a global setting (including the environment) with the intent that the change is temporary, then we do it in a try-finally so that we can ensure that the change is still temporary in the face of exceptions / errors / etc.